src/libraries/System.Text.Json/docs/ParameterizedCtorSpec.md

```diff
@@ -0,0 +1,900 @@
+`PropertyNameCaseInsensitive`. For instance, one could perform performant case-sensitive matching for object properties, and case-insensitive matching for constructor parameters, as needed.
+
+##### Alternatives
+
+1. Do case insensitive matching and don't provide an opt-in for case sensitive matching.
+    
+    - More flexible.
+    - Unnecessary perf hit in the major scenario (`camelCase` parameters, `PascalCase` properties).
+
+2. Do case insensitive matching and provide an opt-in for case sensitive matching.
+
+    - More flexible.
+    - Unnecessary perf hit in the major scenario if option is not specified.
+
+3. Instead of adding the `ParameterCaseInsensitive` option, share the `PropertyCaseInsensitive` option
+
+    - Overloads the term `Property` which is understood to mean object `Property`, not JSON property.
+    - Because the default value is `false`, ioes not work for the major scenario, including for built in types like `Tuple<...>`, without setting the option.
+    - Could degrade performance significantly if turned on in the common scenario, where there are more object properties than constructor parameters.
+
+#### `ParameterNamingPolicy` can be used to facilitate matching
+
+A major scenario enabled by this feature is the round-trippability of "immutable" types that have read-only properties.
+
+For example, given `Point`:
+
+```C#
+public class Point
+{
+    public int XValue { get; }
+
+    public int YValue { get; }
+
+    public Point(int xValue, int yValue) => (XValue, YValue) = (xValue, yValue);
+}
+```
+
+Serializing a new instance of `Point` using a `SnakeCase` policy would yield something like the following:
+
+```C#
+JsonSerializerOptions options = new JsonSerializerOptions
+{
+    PropertyNamingPolicy = new JsonSnakeCaseNamingPolicy()
+};
+
+string serialized = JsonSerializer.Serialize(new Point(1, 2), options);
+Console.WriteLine(serialized); // {"x_value":1,"y_value":2}
+
+Point point = JsonSerializer.Deserialize<Point>(serialized, options);
+Console.WriteLine(point.XValue); // 0
+Console.WriteLine(point.YValue); // 0
+```
+
+If there is no way to specify a `JsonNamingPolicy` for constructor parameters, matching on deserialization
+will be impossible.
+
+This proposal introduces a new `ParameterNamingPolicy` which can be used to facilitate such matching:
+
+```C#
+JsonNamingPolicy policy = new JsonSnakeCaseNamingPolicy(); 
+
+JsonSerializerOptions options = new JsonSerializerOptions
+{
+    ParameterNamingPolicy = policy,
+    PropertyNamingPolicy = policy
+};
+
+string serialized = JsonSerializer.Serialize(new Point(1, 2), options);
+Console.WriteLine(serialized); // {"x_value":1,"y_value":2}
+
+Point point = JsonSerializer.Deserialize<Point>(serialized, options);
+Console.WriteLine(point.XValue); // 1
+Console.WriteLine(point.YValue); // 2
+```
+
+An alternative solution is to share the `PropertyNamingPolicy` used by object properties.
```

Comment:
Here's some conventions I think make sense:
- A ctor argument typically maps 1:1 with a property. This is required for round-tripping the JSON\object. That property likely only has a getter (no setter), but having a setter shouldn't affect this feature since the ctor will still be called.
- A ctor argument is the name of the corresponding property but camel-cased (the property was Pascal-cased).

So consider an approach where we do not treat the object's properties and ctor args as two independent things and that the property and argument naming policies work together.

For example for Serialize it works the same as today:
- The "XValue" properties is serialized as "x_value" based on a snake-casing property naming policy.

And for deserialize it would:
1) Assume the "x_value" JSON is matched to the "XValue" property (already done today).
2) In order to determine if a ctor should be called, the new "argument naming policy" is run (default is camel-case) and converts "XValue" (the property name) to "xValue".
3) The "xValue" is found in the ctor arg (case-sensitive) and then that ctor is called (assuming "yValue" will also be handled like this).

This is nice because it will support `JsonNameAttribute`. If we had a `JsonName["HelloWorld"]` on the "XValue" property, the JSON would be "HelloWorld" (since it ignores the "property naming policy") which would map back to the "XValue" property (like today) which would then be converted to "xValue" argument by the "argument naming policy".

To simplify this, we could just assume the "argument naming policy" is always the camel-casing policy and not expose a way to change that on JsonSerializerOptions. If some edge cases don't work (because camel-casing is not used for argument names) then the workaround is to create a custom converter or create a new type of custom converter that will make this a bit easier (not yet public\designed but we want it for other reasons).

Reply:
For cases when (1) above does not hold because the corresponding property isn't found from the JSON -- then on deserialize we could:

Run the "property naming policy" on the ctor argument names like we do today for properties. Thus we treat argument as "pseudo-properties". Thus "xValue" ctor arg produces "x_value". This would be cached and not done on every deserialization.

For deserialize:
1) The "x_value" JSON did not match up to any properties but did match up to the ctor arg "xValue".
2) Run the "argument naming policy" against "xValue" (which produces "xValue" since it was already camel-cased). So when "property naming policy" is snake-case or camel-case this doesn't really do anything, but would work perfect when no "property naming policy" is specified and we have Pascal-cased JSON.
3) The "x_value" JSON is matched to the "xValue" arg and that ctor is called...

For simplicity we could just do this for all cases (don't try to match up argument name and property name) -- it just means the `JsonName["HelloWorld"]` scenario wouldn't work.

Reply:
I like the proposal to link the property names to the constructor parameter names, as it allows roundtripping scenarios where a `[JsonPropertyName]` is applied to a property, as you bring up.
If there's no property matching the constructor arg, the `PropertyNamingPolicy` in the options will used to determine what JSON to look for.

Reply:
I update the matching semantics as follows:

- Object properties are bound to constructor parameters by converting the properties using  `camelCase` as described in https://github.com/dotnet/runtime/pull/33095#discussion_r387844875. Constructor parameters will then take over the cached JSON property names from the object properties (which will be ignored on deserialization)
- If there's no match for a constructor parameter, the policy specified with `options.PropertyNamingPolicy` will be used to determine what JSON property names to match with.